bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/common/InvocationHandlerAsync.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) 2014,2017 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.internal.common;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.util.function.Consumer;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * Asynchronuous {@link InvocationHandler} implementation.
+ *
+ * Instead of directly invoking the called method it rather queues it in the {@link SafeCallManager} for asynchronous
+ * execution.
+ *
+ * @author Simon Kaufmann - initial contribution and API.
+ *
+ * @param <T>
+ */
+class InvocationHandlerAsync<T> extends AbstractInvocationHandler<T> implements InvocationHandler {
```

Comment:
`@NonNullByDefault`?

Reply:
Due to the yet missing external annotations this leads to some awkward pseudo-inherited annotations and a (in fact) redundant null-check on `method`. I have added it now as you were asking for it, but actually don't think this improved anything really.